docs/Testing.md

```diff
@@ -0,0 +1,127 @@
+# Testing
+
+A test case takes the form of a binary \<basename\>.wasm file, next to that that there will always be a \<basename\>.\<ext\> source file from which the test was originally compiled from which can be used as a reference in the event of an error.
+
+Additionally, any of the following optional auxilary files and directories may be present:
+- \<basename\>.arg
+- \<basename\>.env
+- \<basename\>.dir
+- \<basename\>.stdin
+- \<basename\>.stdout
+- \<basename\>.stderr
+- \<basename\>.status
+
+## Running tests
+
+To run a test do the following steps to prepare for execution:
+
+- Prepare inputs
+  - Given an \<input\> module; take the \<basename\> of said module.
+  - If \<basename\>.\<arg\> exists; take the program arguments from said file.
+  - If \<basename\>.\<env\> exists; take the program environment from said file.
+  - If \<basename\>.\<dir\> exists; preopen the directory from said file.
+  - If \<basename\>.\<stdin\> exists; pipe said file into the program as stdin.
+- Run program
+- Collect results
+  - If \<basename\>.\<stdout\> exists; assert that the programs stdout matches
+    said file.
+  - If \<basename\>.\<stderr\> exists; assert that the programs stdout matches
+    said file.
+  - If \<basename\>.\<status\> exists; assert that the programs stdout matches
+    said file; otherwise assert that the program exited with status 0.
+- Pass
+
+For example:
+
+```bash
+# Usage: $1 <runtime> <path_to_binary.wasm>
+# $1 wasmtime proc_exit-success.wasm
+# $1 wasmer proc_exit-failure.wasm
+#!/usr/bin/env bash
+
+runtime=$1
+input=$2
+
+prefix=${input%.*}
+if [ -e "$prefix.stdin" ]; then
+  stdin="$prefix.stdin"
+else
+  stdin="/dev/null"
+fi
+
+output="$(mktemp -d)"
+stdout_actual="$output/stdout"
+stderr_actual="$output/stderr"
+status_actual="$output/status"
+
+if [ -e "$prefix.arg" ]; then
+  arg=$(cat "$prefix.env")
+else
+  arg=""
+fi
+
+if [ -e "$prefix.env" ]; then
+  env=$(sed -e 's/^/--env /' < "$prefix.env")
+else
+  env=""
+fi
+
+if [ -e "$prefix.dir" ]; then
+  dir="--dir $prefix.dir"
+else
+  dir=""
+fi
+
+status=0
+
+"$runtime" $dir $input -- $arg \
+  < "$stdin" \
+  > "$stdout_actual" \
+  2> "$stderr_actual" \
+  || status=$?
+
+echo $status > "$status_actual"
+
+stdout_expected="$prefix.stdout"
+if [ -e "$stdout_expected" ]; then
+  diff -u "$stderr_expected" "$stderr_actual"
+fi
+
+stderr_expected="$prefix.stderr"
+if [ -e "$stderr_expected" ]; then
+  diff -u "$stdout_expected" "$stdout_actual"
+fi
+
+status_expected="$prefix.status"
+if [ -e "$prefix.status" ]; then
+  diff -u "$status_expected" "$status_actual"
+elif [ ! "$status" -eq "0" ]; then
+  cat $stderr_actual
+  exit 1
+fi
+```
+
+## Writing tests
+
+Any source language may be used to write tests as long as the compiler supports
+the wasm32-unknown-unknown target and is well-known and freely available.
```

Comment:
`wasm32-unknown-unknown` is an llvm-ism so we probably don't want to mention that here.    Perhaps:

"Any source language may be used to write tests as long as the compiler supports
WebAssembly output targeting the WASI API"

The part about "well-known and freely available" will probably want to be tightened up too.

Reply:
> wasm32-unknown-unknown is an llvm-ism so we probably don't want to mention that here. Perhaps:

Targeting wasm32-wasi is probably fine actually.

I just want to make sure we point out that using wasm32-wasi's `main` is generally going to make for a bad test.

For example; trying to test proc_exit in main isn't great as it'll do all the bootstrapping that libc needs but isn't relevant in the context.

> The part about "well-known and freely available" will probably want to be tightened up too.

Yeah I'll need to think about what that means.

I'm thinking:

- We can't allow proprietary compilers.
- We can't allow non-portable compilers.
- We can't allow toy languages that won't be or aren't maintained.

I'll double back to this later.

Reply:
I do tend to think we do want to allow tests to pull in libc. It does call `proc_exit` and some of the stdio routines needlessly pull in `fdstat_get` and perhaps other things, however bare WASI is really inconvenient to code to, so if we have to do everythin in bare WASI, it seems like we'd end with people writing fewer tests. It seems better to have more tests, even if some of them do call more functions than they strictly need to.

Reply:
> I do tend to think we do want to allow tests to pull in libc. It does call proc_exit and some of the stdio routines needlessly pull in fdstat_get and perhaps other things, however bare WASI is really inconvenient to code to, so if we have to do everythin in bare WASI, it seems like we'd end with people writing fewer tests. It seems better to have more tests, even if some of them do call more functions than they strictly need to.

From experience it was a bit tedious to deal with as _main also calls into the prestat, args and env functions as-well (at-least at the time).

Simple to write tests against but early days of WASI in Deno basically couldn't be tested with my libc tests until *everything* was bootstrapped.

I think we'll just have to revisit this once the split is done and it's a bit clearer how that's going to work; but presumably we'd want to generate headers and such based on the the local witx files and not be dependent on the current state of say the wasi-libc implementation and headers or the wasi crate (which would also be a circular dependency).